client/webpack-utils/extract-manifest-plugin.js

```diff
@@ -0,0 +1,124 @@
+const PLUGIN_NAME = 'ExtractManifestPlugin';
+const { ConcatSource } = require( 'webpack-sources' );
+
+class ExtractManifestPlugin {
+	constructor( options = {} ) {
+		this.options = {
+			manifestName: 'manifest',
+			runtimeChunk: 'runtime',
+			globalManifest: 'window.__WEBPACK_MANIFEST',
+			...options,
+		};
+	}
+
+	apply( compiler ) {
+		// Tapping into compiler.make allow us to come after the regular TemplatedPathPlugin
+		compiler.hooks.compilation.tap( PLUGIN_NAME, ( compilation ) => {
+			const globalManifest = this.options.globalManifest;
+			const manifestContent = [ `${ globalManifest }=new Map();` ];
+			const quoteRegex = new RegExp( /^(['"])/ );
+			const variableRegex = new RegExp( /\[(.*?)\]/gi );
+			const variablesProcessed = new Set();
+
+			// This method is hock several times. First, it is called to generate the filename of each chunk. In that
+			// case, `data.hashWithLength` won't be defined and we just forward the output of the original filenameExpression.
+			//
+			// It will also be called to generate an expression that receives a chunkId and returns the filename where that chunkId is
+			// saved, something like `({1:'chunk1', 2:'chunk2'}[chunkId])+".min.js"` (simplified example). In this case, capture the expressions
+			// that do the variable replacement and store them in a global Map. This allows reusing those expressions later in the runtime so we
+			// don't have to concatenate the chunkId map multiple times.
+			//
+			// Simplified example: if filenameExpression is "[name].[chunkhash].js", we end up with
+			//
+			// - In runtime:
+			//		function jsonpScriptSrc(chunkId) {
+			//			return window.__WEBPACK_MANIFEST.get('name')(chunkId) + "." + window.__WEBPACK_MANIFEST.get('chunkhash')(chunkId) + ".js"
+			//		}
+			//
+			// - In manifest:
+			//		window.__WEBPACK_MANIFEST=new Map();
+			//		window.__WEBPACK_MANIFEST.set('name', (chunkId) => {1:'my-name'}[chunkId]);
+			//		window.__WEBPACK_MANIFEST.set('chunkhash', (chunkId) => {1:'abcd1234'}[chunkId]);
+			//
+			// So when webpack calls `jsonpScriptSrc(1)` it gets `my-name.abcd1234.js`
+			compilation.mainTemplate.hooks.assetPath.tap(
+				PLUGIN_NAME,
+				( filenameExpression, options, assetInfo ) => {
+					// When there is no `chunk.hashWithLength`, this hook has been called to generate the file name of a chunk.
+					// Return whatever Webpack generates unmodified.
+					if (
+						options.skipExtractManifestPlugin ||
+						! ( options.chunk && options.chunk.hashWithLength )
+					) {
+						return filenameExpression;
+					}
```

Comment:
I'd split this condition into two, as they don't have anything in common:
1. Bail out when there's a flag that prevents infinite recursion
2. Bail out when the heuristic detects this is a call that we don't want to intercept.